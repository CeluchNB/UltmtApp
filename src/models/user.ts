import { Realm } from '@realm/react'

export const DisplayUserSchema: Realm.ObjectSchema = {
    name: 'DisplayUser',
    embedded: true,
    properties: {
        _id: 'string',
        firstName: 'string',
        lastName: 'string',
        username: 'string',
        localGuest: 'bool?',
    },
}
