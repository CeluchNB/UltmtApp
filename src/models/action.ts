import { DisplayUser } from '../types/user'
import { Realm } from '@realm/react'
import { TeamNumber } from '../types/team'
import { ActionType, Comment, LiveServerActionData } from '../types/action'

export const CommentSchema: Realm.ObjectSchema = {
    name: 'Comment',
    embedded: true,
    properties: {
        user: 'DisplayUser',
        comment: 'string',
        commentNumber: 'int',
    },
}

export class ActionSchema {
    static schema: Realm.ObjectSchema = {
        name: 'Action',
        primaryKey: '_id',
        properties: {
            _id: 'objectId',
            pointId: 'string',
            actionType: 'string',
            actionNumber: 'int',
            playerOne: 'DisplayUser?',
            playerTwo: 'DisplayUser?',
            tags: 'string[]',
            teamNumber: 'string',
            comments: 'Comment[]',
        },
    }

    _id: Realm.BSON.ObjectId
    pointId: string
    actionType: ActionType
    playerOne?: DisplayUser
    playerTwo?: DisplayUser
    tags: string[]
    actionNumber: number
    teamNumber: TeamNumber
    comments: Comment[]

    constructor(action: LiveServerActionData, pointId: string) {
        this._id = new Realm.BSON.ObjectID()
        this.pointId = pointId
        this.actionType = action.actionType
        this.playerOne = action.playerOne
        this.playerTwo = action.playerTwo
        this.tags = action.tags
        this.actionNumber = action.actionNumber
        this.teamNumber = action.teamNumber
        this.comments = action.comments
    }
}
