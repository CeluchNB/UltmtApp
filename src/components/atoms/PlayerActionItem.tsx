import { ActionType } from '../../types/action'
import { Button } from 'react-native-paper'
import { GuestUser } from '../../types/user'
import React from 'react'
import { size } from '../../theme/fonts'
import { useColors } from '../../hooks'
import { StyleSheet, Text, View } from 'react-native'

interface PlayerActionItemProps {
    player: GuestUser
    actions: ('score' | ActionType)[]
    onAction: (action: ActionType | 'score') => void
}

const PlayerActionItem: React.FC<PlayerActionItemProps> = ({
    player,
    actions,
    onAction,
}) => {
    const { colors } = useColors()

    const styles = StyleSheet.create({
        container: {
            display: 'flex',
            flexDirection: 'row',
            marginTop: 10,
            textAlignVertical: true,
        },
        nameContainer: {
            width: '35%',
        },
        name: {
            fontSize: size.fontFifteen,
            color: colors.textPrimary,
        },
        username: {
            fontSize: size.fontSmall,
            color: colors.textPrimary,
        },
        buttonContainer: {
            width: '65%',
        },
        button: {
            borderColor: colors.textPrimary,
            color: colors.textPrimary,
            backgroundColor: colors.primary,
            flex: 1,
            marginRight: 5,
        },
        buttonText: {
            fontSize: size.fontSmall,
        },
    })

    return (
        <View style={styles.container}>
            <View style={styles.nameContainer}>
                <Text style={styles.name} numberOfLines={1}>
                    {player.firstName} {player.lastName}
                </Text>
                {player.username && (
                    <Text style={styles.username}>@{player.username}</Text>
                )}
            </View>
            {actions.map(action => (
                <Button
                    key={action}
                    compact={true}
                    style={styles.button}
                    labelStyle={styles.buttonText}
                    color={colors.textPrimary}
                    collapsable={true}
                    mode="outlined"
                    onPress={() => {
                        onAction(action)
                    }}>
                    {action}
                </Button>
            ))}
        </View>
    )
}

export default PlayerActionItem
