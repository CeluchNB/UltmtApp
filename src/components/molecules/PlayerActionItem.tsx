import { Button } from 'react-native-paper'
import { ClientActionType } from '../../types/action'
import { GuestUser } from '../../types/user'
import PlayerActionTagModal from './PlayerActionTagModal'
import React from 'react'
import { size } from '../../theme/fonts'
import { useColors } from '../../hooks'
import { StyleSheet, Text, View } from 'react-native'

interface PlayerActionItemProps {
    player: GuestUser
    actions: ClientActionType[]
    loading: boolean
    onAction: (
        action: ClientActionType,
        tags: string[],
        player: GuestUser,
    ) => void
}

const PlayerActionItem: React.FC<PlayerActionItemProps> = ({
    player,
    actions,
    loading,
    onAction,
}) => {
    const { colors } = useColors()
    const [modalVisible, setModalVisible] = React.useState(false)
    const [selectedAction, setSelectedAction] = React.useState<
        ClientActionType | undefined
    >()

    const onModalClose = (submit: boolean, tags: string[]) => {
        if (submit && selectedAction) {
            onAction(selectedAction, tags, player)
        }
        setModalVisible(false)
        setSelectedAction(undefined)
    }

    const styles = StyleSheet.create({
        container: {
            display: 'flex',
            flexDirection: 'row',
            marginTop: 10,
            textAlignVertical: true,
        },
        nameContainer: {
            width: '35%',
        },
        name: {
            fontSize: size.fontFifteen,
            color: colors.textPrimary,
        },
        username: {
            fontSize: size.fontSmall,
            color: colors.textPrimary,
        },
        buttonContainer: {
            width: '65%',
        },
        button: {
            borderColor: colors.textPrimary,
            color: colors.textPrimary,
            backgroundColor: colors.primary,
            flex: 1,
            marginRight: 5,
        },
        buttonText: {
            fontSize: size.fontSmall,
        },
    })

    return (
        <View style={styles.container}>
            <View style={styles.nameContainer}>
                <Text style={styles.name} numberOfLines={1}>
                    {player.firstName} {player.lastName}
                </Text>
                {player.username && (
                    <Text style={styles.username}>@{player.username}</Text>
                )}
            </View>
            {actions.map(action => (
                <Button
                    key={action}
                    compact={true}
                    style={styles.button}
                    labelStyle={styles.buttonText}
                    textColor={colors.textPrimary}
                    uppercase={true}
                    collapsable={true}
                    disabled={loading}
                    mode="outlined"
                    onPress={() => {
                        onAction(action, [], player)
                    }}
                    onLongPress={() => {
                        setSelectedAction(action)
                        setModalVisible(true)
                    }}>
                    {action}
                </Button>
            ))}
            <PlayerActionTagModal
                visible={modalVisible}
                onClose={onModalClose}
            />
        </View>
    )
}

export default PlayerActionItem
