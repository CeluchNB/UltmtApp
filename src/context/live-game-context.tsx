import { InGameStatsUser } from '../types/user'
import { TeamNumber } from '../types/team'
import { parseInGameStatsUser } from '../utils/in-game-stats'
import { useFinishGame } from '../hooks/game-edit-actions/use-finish-game'
import { useQuery } from './realm'
import { GameSchema, PointSchema } from '../models'
import React, { ReactNode, createContext, useMemo, useState } from 'react'

interface LiveGameContextData {
    game?: GameSchema
    point?: PointSchema
    team: TeamNumber
    teamId?: string
    players: InGameStatsUser[] | undefined
    tags: string[]
    addTag: (tag: string) => void
    setCurrentPointNumber: (pointNumber: number) => void
    finishGameMutation: {
        finishGame: () => Promise<void>
        finishGameLoading: boolean
        finishGameError?: string
        finishGameReset: () => void
    }
}

export const LiveGameContext = createContext<LiveGameContextData>(
    {} as LiveGameContextData,
)

const LiveGameProvider = ({
    children,
    gameId,
    pointNumber = 1,
    teamNumber = 'one',
}: {
    children: ReactNode
    gameId: string
    pointNumber?: number
    teamNumber?: TeamNumber
}) => {
    // NOT AN IDEAL IMPLEMENTATION - useObject would be better - https://github.com/realm/realm-js/issues/5620
    const game = useQuery<GameSchema>('Game').filtered(`_id == '${gameId}'`)[0]

    const [currentPointNumber, setCurrentPointNumber] = useState(pointNumber)
    const pointQuery = useQuery<PointSchema>(
        {
            type: 'Point',
            query: collection => {
                return collection.filtered(
                    'gameId == $0 AND pointNumber == $1',
                    gameId,
                    currentPointNumber,
                )
            },
        },
        [currentPointNumber],
    )
    const point = useMemo(() => {
        return pointQuery[0]
    }, [pointQuery])

    const [team] = useState<TeamNumber>(teamNumber)
    const teamId = useMemo(() => {
        return team === 'one' ? game?.teamOne._id : game?.teamTwo._id
    }, [team, game])
    const [tags, setTags] = useState(['huck', 'break', 'layout'])

    const addTag = (tag: string) => {
        setTags(curr => [...curr, tag])
    }

    const players = useMemo(() => {
        if (team === 'one') {
            return game?.teamOnePlayers
                .slice()
                .sort((a, b) =>
                    `${a.firstName} ${a.lastName}`.localeCompare(
                        `${b.firstName} ${b.lastName}`,
                    ),
                )
                .map(user => parseInGameStatsUser(user))
        } else if (team === 'two') {
            return game?.teamTwoPlayers
                .slice()
                .sort((a, b) =>
                    `${a.firstName} ${a.lastName}`.localeCompare(
                        `${b.firstName} ${b.lastName}`,
                    ),
                )
                .map(user => parseInGameStatsUser(user))
        }
    }, [game, team])

    const {
        mutateAsync: finishGame,
        isLoading: finishGameLoading,
        error: finishGameError,
        reset: finishGameReset,
    } = useFinishGame(gameId)

    return (
        <LiveGameContext.Provider
            value={{
                game,
                point,
                team,
                players,
                teamId,
                tags,
                addTag,
                setCurrentPointNumber,
                finishGameMutation: {
                    finishGame,
                    finishGameReset,
                    finishGameLoading,
                    finishGameError: finishGameError?.message,
                },
            }}>
            {children}
        </LiveGameContext.Provider>
    )
}

export default LiveGameProvider
