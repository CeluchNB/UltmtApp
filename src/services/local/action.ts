import * as Constants from '../../utils/constants'
import AsyncStorage from '@react-native-async-storage/async-storage'
import { Realm } from '@realm/react'
import { getRealm } from '../../models/realm'
import { throwApiError } from '../../utils/service-utils'
import { ActionSchema, PointSchema } from '../../models'
import { LiveServerAction, SavedServerAction } from '../../types/action'
import { getPointById, savePoint } from './point'

/**
 * Method to save actions locally
 * @param pointId id of point actions belong to
 * @param actions list of action objects
 */
export const saveActions = async (
    pointId: string,
    actions: SavedServerAction[],
) => {
    const pairs: [string, string][] = actions.map(action => [
        `${pointId}:${action._id}`,
        JSON.stringify(action),
    ])
    await AsyncStorage.multiSet(pairs)
}

/**
 * Method to get locally saved actions
 * @param pointId id of point actions belong to
 * @param actionIds list of ids of actions
 * @returns saved user ids or empty if a point is not found
 */
export const getActions = async (
    pointId: string,
    actionIds: string[],
): Promise<SavedServerAction[]> => {
    const keys = actionIds.map(id => `${pointId}:${id}`)
    const actions = await AsyncStorage.multiGet(keys)
    for (const kvp of actions) {
        if (!kvp[1]) {
            return []
        }
    }
    return actions.map(kvp => JSON.parse(kvp[1] || ''))
}

/**
 * Method to remove all locally stored actions from a given point
 * @param pointId point id to remove actions for
 */
export const deleteAllActionsByPoint = async (pointId: string) => {
    const keys = await AsyncStorage.getAllKeys()
    const filteredKeys = keys.filter(key => key.includes(pointId))
    await AsyncStorage.multiRemove(filteredKeys)
}

export const upsertAction = async (
    action: LiveServerAction,
    pointId: string,
): Promise<LiveServerAction> => {
    const realm = await getRealm()
    const point = await realm.objectForPrimaryKey<PointSchema>('Point', pointId)
    if (!point) {
        return throwApiError({}, Constants.GET_POINT_ERROR)
    }
    realm.write(() => {
        const rAction = realm.create<ActionSchema>(
            'Action',
            new ActionSchema(action, pointId),
            Realm.UpdateMode.Modified,
        )
        if (action.teamNumber === 'one') {
            point.teamOneActions = [
                ...new Set([
                    ...point.teamOneActions,
                    rAction._id.toHexString(),
                ]),
            ]
        } else {
            point.teamTwoActions = [
                ...new Set([
                    ...point.teamTwoActions,
                    rAction._id.toHexString(),
                ]),
            ]
        }
    })
    const actions = await realm.objects<ActionSchema>('Action')
    const result = actions.filtered(
        `teamNumber == "${action.teamNumber}" && actionNumber == ${action.actionNumber} && pointId == "${pointId}"`,
    )[0]

    if (!result) {
        return throwApiError({}, Constants.GET_ACTION_ERROR)
    }
    return parseAction(result)
}

export const saveMultipleServerActions = async (
    actions: LiveServerAction[],
    pointId: string,
): Promise<void> => {
    const realm = await getRealm()

    realm.write(() => {
        for (const action of actions) {
            realm.create(
                'Action',
                { ...action, _id: new Realm.BSON.ObjectId(), pointId },
                Realm.UpdateMode.All,
            )
        }
    })
}

export const deleteAction = async (
    teamNumber: 'one' | 'two',
    actionNumber: number,
    pointId: string,
): Promise<LiveServerAction> => {
    const realm = await getRealm()
    const actions = await realm.objects<ActionSchema>('Action')
    const action = actions.filtered(
        `teamNumber == "${teamNumber}" && actionNumber == ${actionNumber} && pointId == "${pointId}"`,
    )[0]

    // remove action from point array
    const point = await getPointById(pointId)
    if (teamNumber === 'one') {
        point.teamOneActions.splice(actionNumber - 1)
    } else {
        point.teamTwoActions.splice(actionNumber - 1)
    }
    await savePoint(point)

    const result = parseLiveAction(action)
    realm.write(() => {
        realm.delete(action)
    })
    return result
}

export const deleteEditableActionsByPoint = async (
    team: 'one' | 'two',
    pointId: string,
): Promise<void> => {
    const realm = await getRealm()
    const actions = await realm
        .objects<ActionSchema>('Action')
        .filtered(`teamNumber == "${team}" && pointId == "${pointId}"`)
    realm.write(() => {
        realm.delete(actions)
    })
}

export const getActionById = async (
    actionId: string,
): Promise<LiveServerAction> => {
    const realm = await getRealm()
    const action = await realm.objectForPrimaryKey<ActionSchema>(
        'Action',
        actionId,
    )
    if (!action) {
        return throwApiError({}, Constants.GET_ACTION_ERROR)
    }
    return parseLiveAction(action)
}

export const getActionsByPoint = async (
    pointId: string,
): Promise<(LiveServerAction & { _id: string; pointId: string })[]> => {
    const realm = await getRealm()
    const actions = await realm
        .objects<ActionSchema>('Action')
        .filtered(`pointId == "${pointId}"`)

    return actions.map(action => {
        return parseAction(action)
    })
}

const parseLiveAction = (schema: ActionSchema): LiveServerAction => {
    return JSON.parse(
        JSON.stringify({
            actionType: schema.actionType,
            actionNumber: schema.actionNumber,
            teamNumber: schema.teamNumber,
            tags: schema.tags,
            comments: schema.comments,
            playerOne: schema.playerOne,
            playerTwo: schema.playerTwo,
        }),
    )
}

const parseAction = (
    schema: ActionSchema,
): LiveServerAction & { _id: string; pointId: string } => {
    return JSON.parse(
        JSON.stringify({
            _id: schema._id.toHexString(),
            pointId: schema.pointId,
            actionType: schema.actionType,
            actionNumber: schema.actionNumber,
            teamNumber: schema.teamNumber,
            tags: schema.tags,
            comments: schema.comments,
            playerOne: schema.playerOne,
            playerTwo: schema.playerTwo,
        }),
    )
}
