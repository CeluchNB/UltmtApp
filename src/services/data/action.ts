import * as Constants from '../../utils/constants'
import EncryptedStorage from 'react-native-encrypted-storage'
import { saveActions as localSaveActions } from '../local/action'
import { addComment as networkAddComment } from '../network/action'
import { throwApiError } from '../../utils/service-utils'
import { withToken } from './auth'
import {
    ClientAction,
    SavedServerAction,
    SubscriptionObject,
} from '../../types/action'
import {
    addComment as networkAddLiveComment,
    createAction as networkCreateAction,
    joinPoint as networkJoinPoint,
    nextPoint as networkNextPoint,
    subscribe as networkSubscribe,
    undoAction as networkUndoAction,
    unsubscribe as networkUnsubscribe,
} from '../network/live-action'

/**
 * Method to join a point
 * @param gameId id of game to join
 * @param pointId id of point to join
 */
export const joinPoint = async (gameId: string, pointId: string) => {
    await networkJoinPoint(gameId, pointId)
}

/**
 * Method to create an action
 * @param action action data
 * @param pointId point action belongs to
 */
export const addAction = async (action: ClientAction, pointId: string) => {
    await networkCreateAction(action, pointId)
}

/**
 * Method to undo an action
 * @param pointId undo last action on point of this id
 */
export const undoAction = async (pointId: string) => {
    await networkUndoAction(pointId)
}

/**
 * Method to add a comment to a live action
 * @param gameId id of game
 * @param pointId id of point
 * @param actionNumber action number
 * @param teamNumber 'one' | 'two' - team reporting action
 * @param comment comment text
 */
export const addLiveComment = async (
    gameId: string,
    pointId: string,
    actionNumber: number,
    teamNumber: 'one' | 'two',
    comment: string,
) => {
    const token = (await EncryptedStorage.getItem('access_token')) || ''
    await networkAddLiveComment(
        token,
        gameId,
        pointId,
        actionNumber,
        teamNumber,
        comment,
    )
}

/**
 * Method to indicate the next point has been started
 * @param pointId id of point that was just finished
 */
export const nextPoint = async (pointId: string) => {
    await networkNextPoint(pointId)
}

/**
 * Method to listen to specific action events
 * @param subscriptions object containing subscription types and methods
 */
export const subscribe = async (subscriptions: SubscriptionObject) => {
    await networkSubscribe(subscriptions)
}

/**
 * Method to unsubscribe from all action websckets
 */
export const unsubscribe = () => {
    networkUnsubscribe()
}

/**
 * Method to add a comment to a saved action.
 * @param actionId id of action
 * @param pointId id of point
 * @param comment comment text
 * @returns updated action
 */
export const addComment = async (
    actionId: string,
    pointId: string,
    comment: string,
): Promise<SavedServerAction> => {
    try {
        const response = await withToken(networkAddComment, actionId, comment)
        const { action } = response.data
        await localSaveActions(pointId, [action])
        return action
    } catch (error) {
        return throwApiError(error, Constants.COMMENT_ERROR)
    }
}
