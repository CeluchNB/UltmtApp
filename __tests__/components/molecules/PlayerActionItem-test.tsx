import { DisplayUser } from '../../../src/types/user'
import { InGameStatsUserFactory } from '../../test-data/user'
import { LiveGameContext } from '../../../src/context/live-game-context'
import PlayerActionItem from '../../../src/components/molecules/PlayerActionItem'
import React from 'react'
import { debounce } from 'lodash'
import { Action, ActionType } from '../../../src/types/action'
import { fireEvent, render, waitFor } from '@testing-library/react-native'

const originalWarn = console.warn.bind(console.warn)
beforeAll(() => {
    console.warn = msg =>
        !msg.toString().includes('flexWrap: `wrap`') && originalWarn(msg)
})
afterAll(() => {
    console.warn = originalWarn
})

const player: DisplayUser = {
    _id: 'user1',
    firstName: 'First',
    lastName: 'Last',
    username: 'firstlast',
}

describe('PlayerActionItem', () => {
    const players = InGameStatsUserFactory.buildList(7)

    const actions: Action[] = [
        {
            action: {
                actionType: ActionType.BLOCK,
                playerOne: player,
                tags: [],
                actionNumber: 1,
                comments: [],
            },
            reporterDisplay: 'Block',
            viewerDisplay: 'Player blocks',
            setPlayersAndUpdateViewerDisplay: jest.fn(),
            setTags: jest.fn(),
        },
        {
            action: {
                actionType: ActionType.CATCH,
                playerOne: player,
                tags: [],
                actionNumber: 1,
                comments: [],
            },
            reporterDisplay: 'Catch',
            viewerDisplay: 'Player catches',
            setPlayersAndUpdateViewerDisplay: jest.fn(),
            setTags: jest.fn(),
        },
        {
            action: {
                actionType: ActionType.TEAM_ONE_SCORE,
                playerOne: player,
                tags: [],
                actionNumber: 1,
                comments: [],
            },
            reporterDisplay: 'score',
            viewerDisplay: 'Player scores',
            setPlayersAndUpdateViewerDisplay: jest.fn(),
            setTags: jest.fn(),
        },
    ]
    beforeAll(() => {
        jest.useFakeTimers({ legacyFakeTimers: true })
    })
    afterAll(() => {
        jest.useRealTimers()
    })

    it('should match snapshot', () => {
        const snapshot = render(
            <LiveGameContext.Provider
                value={{
                    team: 'one',
                    players,
                    tags: ['huck', 'break', 'layout'],
                    addTag: jest.fn(),
                    setCurrentPointNumber: jest.fn(),
                    finishGameMutation: {
                        mutate: jest.fn(),
                        isLoading: false,
                        error: null,
                        reset: jest.fn(),
                    },
                }}>
                <PlayerActionItem
                    player={player}
                    actions={actions}
                    onAction={debounce(jest.fn())}
                    loading={false}
                />
            </LiveGameContext.Provider>,
        )
        expect(snapshot.toJSON()).toMatchSnapshot()
    })

    it('should call action appropriately', async () => {
        const spy = jest.fn()
        const { getByText } = render(
            <LiveGameContext.Provider
                value={{
                    team: 'one',
                    players,
                    tags: ['huck', 'break', 'layout'],
                    addTag: jest.fn(),
                    setCurrentPointNumber: jest.fn(),
                    finishGameMutation: {
                        mutate: jest.fn(),
                        isLoading: false,
                        error: null,
                        reset: jest.fn(),
                    },
                }}>
                <PlayerActionItem
                    player={player}
                    actions={actions}
                    onAction={debounce(spy)}
                    loading={false}
                />
            </LiveGameContext.Provider>,
        )

        const scoreBtn = getByText('score')
        fireEvent.press(scoreBtn)
        await waitFor(() => {
            expect(spy).toHaveBeenCalledWith(actions[2])
        })
    })

    it('should handle tag call', async () => {
        const spy = jest.fn()
        const { getByText } = render(
            <LiveGameContext.Provider
                value={{
                    team: 'one',
                    players,
                    tags: ['huck', 'break', 'layout'],
                    addTag: jest.fn(),
                    setCurrentPointNumber: jest.fn(),
                    finishGameMutation: {
                        mutate: jest.fn(),
                        isLoading: false,
                        error: null,
                        reset: jest.fn(),
                    },
                }}>
                <PlayerActionItem
                    player={player}
                    actions={actions}
                    onAction={debounce(spy)}
                    loading={false}
                />
            </LiveGameContext.Provider>,
        )

        const scoreBtn = getByText('score')
        fireEvent(scoreBtn, 'onLongPress')

        const huckBtn = getByText('huck')
        fireEvent.press(huckBtn)

        const doneBtn = getByText('done')
        fireEvent.press(doneBtn)

        expect(actions[2].setTags).toHaveBeenCalledWith(['huck'])

        await waitFor(() => {
            expect(spy).toHaveBeenCalledWith({
                ...actions[2],
            })
        })
    })
})
