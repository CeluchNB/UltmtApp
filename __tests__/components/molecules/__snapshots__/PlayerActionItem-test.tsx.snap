// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerActionItem should match snapshot 1`] = `
<View
  style={
    {
      "display": "flex",
      "flexDirection": "row",
      "marginTop": 10,
    }
  }
>
  <View
    style={
      {
        "width": "35%",
      }
    }
  >
    <Text
      numberOfLines={1}
      style={
        {
          "color": "#ffd600",
          "fontSize": 30.821917808219176,
        }
      }
    >
      First
       
      Last
    </Text>
    <Text
      style={
        {
          "color": "#ffd600",
          "fontSize": 20.54794520547945,
        }
      }
    >
      @
      firstlast
    </Text>
  </View>
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#303030",
          "borderRadius": 20,
          "flex": 1,
          "marginRight": 5,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
          "shadowRadius": 0,
          "width": "100%",
        }
      }
      testID="button-container-outer-layer"
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "#303030",
            "borderColor": "#ffd600",
            "borderRadius": 20,
            "borderStyle": "solid",
            "borderWidth": 1,
            "color": "#ffd600",
            "flex": undefined,
            "minWidth": "auto",
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
          }
        }
        testID="button-container"
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "overflow": "hidden",
              },
              {
                "borderRadius": 20,
              },
            ]
          }
          testID="button"
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                [
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "rgba(28, 27, 31, 1)",
                    "writingDirection": "ltr",
                  },
                  [
                    {
                      "fontFamily": "System",
                      "fontSize": 14,
                      "fontWeight": "500",
                      "letterSpacing": 0.1,
                      "lineHeight": 20,
                    },
                    [
                      {
                        "marginHorizontal": 16,
                        "marginVertical": 9,
                        "textAlign": "center",
                      },
                      false,
                      {
                        "marginHorizontal": 24,
                        "marginVertical": 10,
                      },
                      {
                        "marginHorizontal": 8,
                      },
                      {
                        "textTransform": "uppercase",
                      },
                      {
                        "color": "#ffd600",
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0.1,
                        "lineHeight": 20,
                      },
                      {
                        "fontSize": 20.54794520547945,
                      },
                    ],
                  ],
                ]
              }
              testID="button-text"
            >
              Block
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#303030",
          "borderRadius": 20,
          "flex": 1,
          "marginRight": 5,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
          "shadowRadius": 0,
          "width": "100%",
        }
      }
      testID="button-container-outer-layer"
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "#303030",
            "borderColor": "#ffd600",
            "borderRadius": 20,
            "borderStyle": "solid",
            "borderWidth": 1,
            "color": "#ffd600",
            "flex": undefined,
            "minWidth": "auto",
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
          }
        }
        testID="button-container"
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "overflow": "hidden",
              },
              {
                "borderRadius": 20,
              },
            ]
          }
          testID="button"
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                [
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "rgba(28, 27, 31, 1)",
                    "writingDirection": "ltr",
                  },
                  [
                    {
                      "fontFamily": "System",
                      "fontSize": 14,
                      "fontWeight": "500",
                      "letterSpacing": 0.1,
                      "lineHeight": 20,
                    },
                    [
                      {
                        "marginHorizontal": 16,
                        "marginVertical": 9,
                        "textAlign": "center",
                      },
                      false,
                      {
                        "marginHorizontal": 24,
                        "marginVertical": 10,
                      },
                      {
                        "marginHorizontal": 8,
                      },
                      {
                        "textTransform": "uppercase",
                      },
                      {
                        "color": "#ffd600",
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0.1,
                        "lineHeight": 20,
                      },
                      {
                        "fontSize": 20.54794520547945,
                      },
                    ],
                  ],
                ]
              }
              testID="button-text"
            >
              Catch
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#303030",
          "borderRadius": 20,
          "flex": 1,
          "marginRight": 5,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
          "shadowRadius": 0,
          "width": "100%",
        }
      }
      testID="button-container-outer-layer"
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "#303030",
            "borderColor": "#ffd600",
            "borderRadius": 20,
            "borderStyle": "solid",
            "borderWidth": 1,
            "color": "#ffd600",
            "flex": undefined,
            "minWidth": "auto",
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
          }
        }
        testID="button-container"
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "overflow": "hidden",
              },
              {
                "borderRadius": 20,
              },
            ]
          }
          testID="button"
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                [
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "rgba(28, 27, 31, 1)",
                    "writingDirection": "ltr",
                  },
                  [
                    {
                      "fontFamily": "System",
                      "fontSize": 14,
                      "fontWeight": "500",
                      "letterSpacing": 0.1,
                      "lineHeight": 20,
                    },
                    [
                      {
                        "marginHorizontal": 16,
                        "marginVertical": 9,
                        "textAlign": "center",
                      },
                      false,
                      {
                        "marginHorizontal": 24,
                        "marginVertical": 10,
                      },
                      {
                        "marginHorizontal": 8,
                      },
                      {
                        "textTransform": "uppercase",
                      },
                      {
                        "color": "#ffd600",
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0.1,
                        "lineHeight": 20,
                      },
                      {
                        "fontSize": 20.54794520547945,
                      },
                    ],
                  ],
                ]
              }
              testID="button-text"
            >
              score
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    testID="base-modal"
    transparent={true}
    visible={false}
  />
</View>
`;
